llvm/lib/Transforms/UncoalescedAnalysis/UncoalescedAnalysis.cpp

```diff
@@ -0,0 +1,419 @@
+    const Value* in1 = BO->getOperand(0);
+    const Value* in2 = BO->getOperand(1);
+    // Get values
+    MultiplierValue v1, v2;
+    v1 = st.getValue(in1);
+    v2 = st.getValue(in2);
+    // Apply operation to get the resultant value.
+    MultiplierValue v;
+    auto op = BO->getOpcode();
+    switch (op) {
+      case Instruction::URem:
+      case Instruction::SRem:
+      case Instruction::AShr:
+      case Instruction::LShr:
+        v = v1;
+        break;
+      case Instruction::Add:
+        v = v1 + v2;
+        break;
+      case Instruction::Sub:
+        v = v1 + (- v2);
+        break;
+      case Instruction::Shl:
+      case Instruction::Mul:
+        v = v1 * v2;
+        break;
+      case Instruction::UDiv:
+      case Instruction::SDiv:
+        v = v1 * v2;
+        break;
+      case Instruction::Or:
+        v = v1 || v2;
+        break;
+      case Instruction::And:
+        v = v1 && v2;
+        break;
+      case Instruction::Xor:
+        v = (v1 && (-v2)) || (v2 && (-v1));
+        break;
+      default:
+        v = MultiplierValue(TOP);
+        break;
+    }
+    st.setValue(BO, v);
+
+  } else if (isa<CastInst>(I)) {
+    st.setValue(I, st.getValue(I->getOperand(0)));
+
+  } else if (isa<CallInst>(I)) {
+    const CallInst *CI = cast<CallInst>(I);
+    if (CI->isInlineAsm()) {
+      st.setValue(CI, MultiplierValue(TOP));
+    } else {
+      // If function has no name, return!
+      Function *calledF = CI->getCalledFunction();
+      if (!calledF->hasName()) {
+        st.setValue(CI, MultiplierValue(TOP));
+      } else {
+        std::string name = demangle(calledF->getName().str());
+        if (!name.compare("__HIP_Coordinates<__HIP_ThreadIdx>::__X::operator unsigned int() const")) {
+          st.setValue(CI, MultiplierValue(ONE));
+        } else if (
+            !name.compare("__HIP_Coordinates<__HIP_ThreadIdx>::__Y::operator unsigned int() const") ||
+            !name.compare("__HIP_Coordinates<__HIP_ThreadIdx>::__Z::operator unsigned int() const") ||
+            !name.compare("__HIP_Coordinates<__HIP_BlockDim>::__X::operator unsigned int() const") ||
+            !name.compare("__HIP_Coordinates<__HIP_BlockDim>::__Z::operator unsigned int() const") ||
+            !name.compare("__HIP_Coordinates<__HIP_BlockDim>::__Y::operator unsigned int() const") ||
+            !name.compare("__HIP_Coordinates<__HIP_BlockIdx>::__X::operator unsigned int() const") ||
+            !name.compare("__HIP_Coordinates<__HIP_BlockIdx>::__Y::operator unsigned int() const") ||
+            !name.compare("__HIP_Coordinates<__HIP_BlockIdx>::__Z::operator unsigned int() const") ||
+            !name.compare("__HIP_Coordinates<__HIP_GridDim>::__X::operator unsigned int() const") ||
+            !name.compare("__HIP_Coordinates<__HIP_GridDim>::__Y::operator unsigned int() const") ||
+            !name.compare("__HIP_Coordinates<__HIP_GridDim>::__Z::operator unsigned int() const")
```

Comment:
```suggestion
            std::regex_match(name, std::regex{ "__HIP_Coordinates<__HIP_ThreadIdx>::__(Y|Z)::operator unsigned int\(\) const" } ||
            std::regex_match(name, std::regex{ "__HIP_Coordinates<__HIP_(BlockDim|BlockIdx|GridDim)>::__(X|Y|Z)::operator unsigned int\(\) const" })
```

This style of logic is bit confusing, since at first glance, all the `!` suggested to me, you _don't_ want to match to these strings. If you don't want to go the regex route, you should probably use ` == 0` -- in fact, clang-tidy will suggest that via the `readability-implicit-bool-conversion` option.

Also, you could technically condense the regex to:

```cpp
std::regex_match(name, std::regex{ "__HIP_Coordinates<__HIP_(ThreadIdx|BlockDim|BlockIdx|GridDim)>::__(X|Y|Z)::operator unsigned int\(\) const" })
```

since `"__HIP_Coordinates<__HIP_ThreadIdx>::__X::operator unsigned int() const"` will be caught by the preceding condition.